### PR TITLE
Add instrument compiler flag

### PR DESCRIPTION
**Pull request by drognanar**
Adds an instrument compiler flag that enables instrumenting the compilation to allow dynamic analysis.

### CODE REVIEW

**Comment by drognanar** [comment]
FYI @dotnet/roslyn-interactive

**Comment by ManishJayaswal** [comment]
@jcouv @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisResourceTests.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ public static void FlushPayload()
 ";
 
             var c = CreateCompilationWithMscorlib(Parse(source, @"C:\myproject\doc1.cs"));
-            var peImage = c.EmitToArray(EmitOptions.Default.WithEmitDynamicAnalysisData(true));
+            var peImage = c.EmitToArray(EmitOptions.Default.WithInstrument("Test.Flag"));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisResourceTests.cs]: Should we have a test to verify that instrumentation spans are not added to the IL when the "instrument" flag is not present?

Comment (amcasey) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisResourceTests.cs]: Yes, that's quite an important thing to test.

**Comment by amcasey** [comment]
ETA: `Feature branch? If yes, then this failure is expected - we do not currently support running internal jobs against feature branches.`

File: src/Compilers/Core/CodeAnalysisTest/Emit/EmitOptionsTests.cs [hunk 2]
```diff
@@ -66,18 +66,19 @@ public void TestFieldsForEqualsAndGetHashCode()
...
-                "FileAlignment",
-                "HighEntropyVirtualAddressSpace",
-                "BaseAddress",
-                "DebugInformationFormat",
-                "OutputNameOverride",
-                "PdbFilePath",
-                "RuntimeMetadataVersion",
-                "TolerateErrors",
-                "IncludePrivateMembers");
+                nameof(EmitOptions.EmitDynamicAnalysisData),
```
Comment (amcasey) [inline src/Compilers/Core/CodeAnalysisTest/Emit/EmitOptionsTests.cs]: Thanks!

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 3]
```diff
@@ -89,6 +91,35 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
         /// </summary>
         public string RuntimeMetadataVersion { get; private set; }
 
+        // 1.2 BACKCOMPAT OVERLOAD -- DO NOT TOUCH
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Isn't the backcompat already handled by the fact that the new `string instrument` parameter to the constructor is optional?

**Comment by amcasey** [comment]
:+1:, once you add the test @jcouv asked for.

**Comment by JohnHamby** [comment]
I sign off.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 3 above)
Comment (drognanar) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: I followed the guidelines outlined in: https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md

**Comment by drognanar** [comment]
Added the test.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 3 above)
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Thanks

**Comment by jcouv** [comment]
LGTM

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 4]
```diff
@@ -2678,6 +2679,27 @@ public void ParseOut2()
             Assert.Null(parsedArgs.CompilationOptions.ModuleName);
         }
 
+        [Fact]
+        public void ParseInstrumentTestNames()
+        {
+            var parsedArgs = DefaultParse(SpecializedCollections.EmptyEnumerable<string>(), _baseDirectory);
+            Assert.Equal("", parsedArgs.EmitOptions.Instrument);
+
+            parsedArgs = DefaultParse(new[] { @"/instrument:", "a.cs" }, _baseDirectory);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Please add test with ```@"/instrument"```

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 5]
```diff
@@ -2678,6 +2679,27 @@ public void ParseOut2()
...
+            parsedArgs.Errors.Verify(
+                // error CS2006: Command-line syntax error: Missing '<text>' for 'instrument' option
+                Diagnostic(ErrorCode.ERR_SwitchNeedsString).WithArguments("<text>", "instrument"));
+
+            parsedArgs = DefaultParse(new[] { "/instrument:", "Test.Flag.Name", "a.cs" }, _baseDirectory);
+            parsedArgs.Errors.Verify(
+                // error CS2006: Command-line syntax error: Missing '<text>' for 'instrument' option
+                Diagnostic(ErrorCode.ERR_SwitchNeedsString).WithArguments("<text>", "instrument"));
+
+            parsedArgs = DefaultParse(new[] { "/instrument:Test.Flag.Name", "a.cs" }, _baseDirectory);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Please add test with quoted string as a value with space in the middle.

File: src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs [hunk 6]
```diff
@@ -60,6 +60,12 @@ public ITaskItem[] Analyzers
         // We do not support BugReport because it always requires user interaction,
         // which will cause a hang.
 
+        public string Instrument
```
Comment (AlekseyTs) [inline src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs]: Please add doc comment.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 7]
```diff
@@ -35,6 +32,11 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
         public bool IncludePrivateMembers { get; private set; }
 
         /// <summary>
+        /// 
+        /// </summary>
+        public string Instrument { get; private set; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Please provide a comment.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 8]
```diff
@@ -347,5 +375,10 @@ public EmitOptions WithIncludePrivateMembers(bool value)
 
             return new EmitOptions(this) { IncludePrivateMembers = value };
         }
+
+        public EmitOptions WithInstrument(string instrument)
+        {
+            return new EmitOptions(this) { Instrument = instrument };
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Please follow pattern for other similar methods, i. e. check if we can simply return ```this```.

File: src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb [hunk 9]
```diff
@@ -537,6 +538,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                             sdkPaths.AddRange(ParseSeparatedPaths(value))
                             Continue For
 
+                        Case "instrument"
+                            If String.IsNullOrEmpty(value) Then
+                                AddDiagnostic(diagnostics, ERRID.ERR_ArgumentRequired, "instrument", ":<types>")
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb]: Why does it say "types"?

File: src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb [hunk 10]
```diff
@@ -537,6 +538,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                            End If
+
+                            instrument = value
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCommandLineParser.vb]: Should we do this if we reported an error above?

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 11]
```diff
@@ -2678,6 +2679,27 @@ public void ParseOut2()
...
+            parsedArgs.Errors.Verify();
+            Assert.Equal("Test.Flag.Name", parsedArgs.EmitOptions.Instrument);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Please assert this property for error cases too.

File: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb [hunk 12]
```diff
@@ -416,6 +416,24 @@ a.vb
         End Sub
 
         <Fact>
+        Public Sub ParseInstrumentTestNames()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb]: Comments for C# tests are applicable here as well.

**Pull request merged**